### ISSUE TITLE
Duplicate touch events on Android

### ISSUE THREAD

**Issue by kollivier** [photonstorm/phaser-ce#32]
When switching to v2.7.3, I have noticed strange behavior on Android where each touch event would trigger twice. Doing some debugging, I discovered that both the mspointer and touch device pointer types were being registered, and both would fire their own event when a touch event occurred.

I'm not sure why I did not encounter this in previous versions, but I suspect it has something to do with the move towards using the W3C version of PointerEvents rather than the old msPointer API:

https://github.com/photonstorm/phaser-ce/commit/6caec0d14d96a5974cff0d31f7daa885f7a08b13

Both the MSPointer class docs, as well as the code, suggest this is intended for use on Windows devices only, but window.PointerEvent is supported cross-platform, at least in Chrome. I'm not too familiar with the API, but I suspect the old window.navigator.(ms)PointerEnabled variables were returning true only on Windows devices, even on Chrome, which would explain why this problem didn't occur until the commit linked above. I also don't understand why I don't see double events on desktop Chrome, as it should be registering the pointer there as well...

As such, I'm not sure of the proper fix. One way would be to ensure that Phaser only ever registers one touch device handler, but if so, which should it prioritize? Another option would be to do a Windows OS check before setting mspointer, with a note that a better fix will come at a later date. This would adhere to the documented behavior of MSPointer and avoid any confusion caused by MSPointer being used on platforms other than Windows. (I would guess Phaser 3 would be the place to do any serious refactoring to make PointerEvent truly a cross-platform input option.)

**Comment by NeoHoffa** [comment]
Has there been any work around for this yet? I'm running into the same issue. I need multitouch enabled on Android, all of my buttons are getting triggered twice on Android.

**Comment by kenfoo** [comment]
Temporary hackish workaround on my side was to simply disable touch if both PointerEvent is present.

        if (typeof PointerEvent == 'function') {
            this.game.input.touch.enabled = false;
        }

Phaser-ce probably should be modified to pick only one of either PointerEvent or TouchEvent

**Comment by NeoHoffa** [comment]
The latest version (2.7.5) seems to bring this issue back. The issue also involves when being run within an iframe on Android. I implemented a work around that did work with v2.7.3 which uses a timer to reset a boolean so it's not triggered twice. It was working great, but switching over to v2.7.5, brought back the double events.

**Comment by kenfoo** [comment]
Could you share a simple sample project / code that reproduces this issue, and also the workaround that you wrote for 2.7.3 ? Was the workaround done on your own code, or was done to the phaser-ce 2.7.3 code itself?

**Comment by NeoHoffa** [comment]
@kenfoo, It was done within my own code. I added a variable to act as a flag, that is reset after a short time. I placed the reset code within a function so I could use it for multiple buttons.

```
this.pressIsAllowed = true;

function resetAllowedPress(){
    this.pressIsAllowed = false;
    NeoGame.game.time.events.add(100, function() {
        this.pressIsAllowed = true;
    }, this);
}

var helpButton = this.game.add.button(0, -80, "ui-assets", function(){

    if(this.pressIsAllowed){
        resetAllowedPress();
        
        // do stuff here

    }
}, this, "help-button-over", "help-button");

```

However, doing this no longer works in version 2.7.5.

**Comment by goldfire** [comment]
Has anything been figured out on this? We are facing the same issue.

**Comment by NeoHoffa** [comment]
I needed to use 2.7.3, and the workaround I mentioned above. However, I haven't looked into this issue since 2.7.5, and it is up to 2.7.10 now. Maybe checkout the change log to see if it has been addressed.

**Comment by goldfire** [comment]
@NeoHoffa We are on 2.7.10 and still having duplicated events.

**Comment by NeoHoffa** [comment]
@goldfire Have you tried the workaround I noted above? That is what worked for me.

**Comment by goldfire** [comment]
Yes, it technically works, but having to do this everywhere we use input events is a huge pain and shouldn't be necessary. Leads to some very ugly and unmaintainable code when you have input events all over the place.

**Comment by NeoHoffa** [comment]
@goldfire Agreed! Since this issue is closed, perhaps the conversation should continue on issue #71, which is when this issue was closed.

**Comment by SBCGames** [comment]
Hi, I can confirm that this issue is still there when game is in iframe.

Here is source for minimal example - two files:
**index.html** - contains iframe
```html
<html>
  <head>
    <meta charset="utf-8" />
    <title>Parent</title>
  </head>
  <body>
    <iframe name="Game" src="game.html" width="1280" height="800" frameborder="0"></iframe>
  </body>
</html>
```

and **game.html** - game itself:
```html
<!DOCTYPE html>
<html lang="ja">
<head>
    <title>Game</title>
    <meta charset="utf-8">

    <script src="phaser.min.js"></script>
</head>
<body>
    <div id="content"></div>
    <div id="orientation"></div>
    
    <script>
      var game = new Phaser.Game(800, 600, Phaser.AUTO, 'phaser-example', { create: create });

      function create() {

        var graphics = game.add.graphics(0, 0);

        graphics.beginFill(0xFF0000, 1);
        graphics.drawCircle(300, 300, 100);
        
        graphics.inputEnabled = true;
        graphics.events.onInputUp.add(function () {
          console.log("Clicked...");
        }, this);
    }
    
    </script>
</body>
</html>
```

Of course, put **phaser.min.js** into the same directory.

Now, when running this even in PC Chrome with device toolbar toggled on, you can click red circle and see two console outputs - no matter what device you select (see, I have iPad on screenshots). Two events are fired - onMouseUp and onPointerUp:
![pointerup](https://cloud.githubusercontent.com/assets/8962889/26804094/1015327e-4a47-11e7-948b-b27af8822fa8.jpg)
![mouseup](https://cloud.githubusercontent.com/assets/8962889/26804095/101a3e22-4a47-11e7-8516-ebb911f44018.jpg)

Only happens if in iframe! If you open just game.html, then event is fired only once (onPointerUp).

Also, in Phaser 2.7.3 iframe works fine - only one event, but in 2.7.5 it fires two events.

**Comment by SBCGames** [comment]
changing this:
```javascript
        this.mouse.start();
        if (!this.game.device.mspointer)
        {
            // Chrome >= 55 fires both PointerEvent and TouchEvent.
            // Pick only one.
            this.touch.start();
        }
        this.mspointer.start();
        this.mousePointer.active = true;
```
to this:
```javascript
        this.mouse.start();
        // if (!this.game.device.mspointer)
        // {
            // Chrome >= 55 fires both PointerEvent and TouchEvent.
            // Pick only one.
        //    this.touch.start();
        // }
        this.touch.start();
        if (this.game.device.mspointer) { this.mspointer.start(); }
        this.mousePointer.active = true;
```

worked for me - now I get only one event in iframe and also when game is run outside of iframe. Not sure, why it works... I did not check all the classes deeply.

**Comment by kenfoo** [comment]
I tried this with the latest Phaser-ce 2.8.0. I couldn't replicate it on Chrome for desktop. (Chrome/58.0.3029.110 )

But, I did manage to replicate the issue of double events on Chrome on Android (chrome 58.0.3029.83 Mobile). 

Using the desktop Chrome's remote device inspector, I see that in addition to the pointer event, a mouse event was also triggered. (mouseup), causing the second click.

Not sure what is the best solution for this at the moment. I think the solution requires some thought though, as based on the w3c definition, a pointer event can occur for ANY touch, pen or mouse event.

Perhaps allow the double event, but leave it up to the application detect the type of event based on pointerType of the pointer event?
See: https://www.w3.org/TR/pointerevents/#widl-PointerEvent-pointerType

I'm rather busy at the moment though, perhaps someone else could roll up their sleeve and work on this?

**Comment by samme** [comment]
- <https://codepen.io/samme/details/BZNZgG/> (frame)
- <https://codepen.io/samme/debug/BZNZgG> (no frame)

Possible workaround is

```javascript
game.input.mouse.enabled = !game.device.mspointer
```

**Comment by netgfx** [comment]
Is this issue fixed on 2.8.0?

**Comment by samme** [comment]
> Is this issue fixed on 2.8.0?

No.

**Comment by bobhfut** [comment]
I also meet this question , does it solved in new version??

**Comment by samme** [comment]
No, this issue is still open.

**Comment by RollinSafary** [comment]
try turning off mouse pointer for touch devices, this helped me,
you need do that in your bootstate,
`
function:init(){
game.input.maxPointers = 1;
if(game.device.touch){
game.input.mouse.stop()
}
}

`

double event firing is result of touch and mouse pointer events, when you touch on android devices, it handle the event for mousePointer and for touchPointer, so fires twice,
hope this will help you

**Comment by samme** [comment]
https://github.com/photonstorm/phaser-ce/blob/39e30c8567b61ca204cb7a62bde6ecdee6bfec24/src/input/Mouse.js#L218-L222

**Comment by samme** [comment]
Expect v2.11.0 resolved this, please reopen if not.

**Comment by KellyMAM** [comment]
Is anyone else having the same problem in v3.55.2?

**Comment by KellyMAM** [comment]
Same issue as others where an emulator was triggering mouse and touch.

Updated quick fix for v3.55.2
```
if (game.device.input.touch)
  {
    game.input.mouse.enabled = false
  }
```
